File: src/lib/mongo/mongoshAst/lexer.ts
// Sprint 384 (2026-05-17) — mongosh AST tokenizer + token-stream helpers.
//
// 작성 이유: sprint-383 의 1033 LOC `mongoshAst.ts` 가 max-lines soft cap
// (500) 을 두 배 초과했다. lexer (이 파일) / parser (`parser.ts`) /
// argList (`argList.ts`) 3개 + public re-export (`index.ts`) 4-파일로
// split — behavior 변경 0. 본 파일은 tokenizer + `Token` shape + `TokenStream`
// 만 export 한다. error 헬퍼 (`err`, `tokErr`, `describeToken`) 와
// public 결과 타입은 lexer / parser 양쪽에서 공유되므로 여기 둔다.

// ---------------------------------------------------------------------------
// Public API — result types (shared by parser; surfaced via index.ts)
// ---------------------------------------------------------------------------

export type MongoshErrorKind =
  | "unsupported-syntax"
  | "bson-literal"
  | "multiple-statements"
  | "variable-declaration"
  | "function-declaration"
  | "non-db-statement";

export interface MongoshAdminCommand {
  readonly kind: "admin-command";
  /** `runCommand` 또는 `adminCommand`. caller 가 dispatch 분기에 사용. */
  readonly commandName: "runCommand" | "adminCommand";
  /** 본문 (`{<command>: <arg>, ...options}`). JSON-compatible. */
  readonly body: Record<string, unknown>;
}

export interface MongoshCollectionCommand {
  readonly kind: "collection-command";
  readonly collection: string;
  readonly method: string;
  readonly args: readonly unknown[];
}

export interface MongoshParseError {
  readonly kind: "error";
  readonly errorKind: MongoshErrorKind;
  readonly message: string;
}

export type MongoshStatementResult =
  | MongoshAdminCommand
  | MongoshCollectionCommand
  | MongoshParseError;

// ---------------------------------------------------------------------------
// Internal — token shape
// ---------------------------------------------------------------------------

export type Token =
  | { kind: "ident"; value: string; pos: number }
  | { kind: "string"; value: string; pos: number }
  | { kind: "number"; value: number; pos: number }
  | { kind: "punct"; value: string; pos: number };

export interface TokenizeOk {
  readonly kind: "ok";
  readonly tokens: Token[];
}
export interface TokenizeErr {
  readonly kind: "error";
  readonly error: MongoshParseError;
}

const PUNCT_CHARS = new Set(["{", "}", "[", "]", "(", ")", ",", ":", ";", "."]);

// ---------------------------------------------------------------------------
// Internal — tokenizer
// ---------------------------------------------------------------------------

export function tokenize(src: string): TokenizeOk | TokenizeErr {
  const tokens: Token[] = [];
  let i = 0;
  while (i < src.length) {
    const ch = src[i]!;

    if (ch === " " || ch === "\t" || ch === "\n" || ch === "\r") {
      i++;
      continue;
    }

    // Line comment `// ... \n`
    if (ch === "/" && src[i + 1] === "/") {
      while (i < src.length && src[i] !== "\n") i++;
      continue;
    }
    // Block comment `/* ... */` — nested NOT supported (mongosh parity):
    // the lexer closes at the first `*/`, leaving any trailing `c */` as
    // garbage that the parser will then choke on.
    if (ch === "/" && src[i + 1] === "*") {
      i += 2;
      while (i < src.length && !(src[i] === "*" && src[i + 1] === "/")) i++;
      if (i >= src.length) {
        return tokErr("unterminated /* ... */ comment");
      }
      i += 2;
      continue;
    }

    // Arrow function `=>` — explicit reject
    if (ch === "=" && src[i + 1] === ">") {
      return tokErr(
        "arrow functions (`=>`) are not supported — callbacks are outside scope",
      );
    }

    // Template literal — interpolation-free is treated as a plain string
    // (sprint-383). `${` inside the template rejects.
    if (ch === "`") {
      const result = readTemplate(src, i);
      if (result.kind === "error")
        return { kind: "error", error: result.error };
      tokens.push({ kind: "string", value: result.value, pos: i });
      i = result.next;
      continue;
    }

    if (ch === '"' || ch === "'") {
      const result = readString(src, i, ch);
      if (result.kind === "error")
        return { kind: "error", error: result.error };
      tokens.push({ kind: "string", value: result.value, pos: i });
      i = result.next;
      continue;
    }

    // Number literal — leading `-` only if followed by a digit
    if (ch === "-" && /[0-9]/.test(src[i + 1] ?? "")) {
      const result = readNumber(src, i);
      tokens.push({ kind: "number", value: result.value, pos: i });
      i = result.next;
      continue;
    }
    if (/[0-9]/.test(ch)) {
      const result = readNumber(src, i);
      tokens.push({ kind: "number", value: result.value, pos: i });
      i = result.next;
      continue;
    }

    // Identifier (incl. `$`-prefixed)
    if (/[A-Za-z_$]/.test(ch)) {
      let j = i + 1;
      while (j < src.length && /[A-Za-z0-9_$]/.test(src[j]!)) j++;
      tokens.push({ kind: "ident", value: src.slice(i, j), pos: i });
      i = j;
      continue;
    }

    if (PUNCT_CHARS.has(ch)) {
      tokens.push({ kind: "punct", value: ch, pos: i });
      i++;
      continue;
    }

    return tokErr(`unexpected character \`${ch}\` at offset ${i}`);
  }

  return { kind: "ok", tokens };
}

function readString(
  src: string,
  start: number,
  quote: string,
):
  | { kind: "ok"; value: string; next: number }
  | { kind: "error"; error: MongoshParseError } {
  let i = start + 1;
  let value = "";
  while (i < src.length) {
    const ch = src[i]!;
    if (ch === "\\") {
      const next = src[i + 1];
      if (next === undefined) {
        return {
          kind: "error",
          error: err("unsupported-syntax", "unterminated string escape"),
        };
      }
      if (next === "u") {
        const hex = src.slice(i + 2, i + 6);
        if (!/^[0-9a-fA-F]{4}$/.test(hex)) {
          return {
            kind: "error",
            error: err("unsupported-syntax", "invalid \\u escape in string"),
          };
        }
        value += String.fromCharCode(parseInt(hex, 16));
        i += 6;
        continue;
      }
      const ESC: Record<string, string> = {
        n: "\n",
        r: "\r",
        t: "\t",
        b: "\b",
        f: "\f",
        "\\": "\\",
        "'": "'",
        '"': '"',
        "/": "/",
      };
      value += ESC[next] ?? next;
      i += 2;
      continue;
    }
    if (ch === quote) {
      return { kind: "ok", value, next: i + 1 };
    }
    value += ch;
    i++;
  }
  return {
    kind: "error",
    error: err("unsupported-syntax", "unterminated string literal"),
  };
}

function readTemplate(
  src: string,
  start: number,
):
  | { kind: "ok"; value: string; next: number }
  | { kind: "error"; error: MongoshParseError } {
  let i = start + 1;
  let value = "";
  while (i < src.length) {
    const ch = src[i]!;
    if (ch === "\\") {
      // Allow common escapes; unknown escapes pass through.
      const next = src[i + 1];
      if (next === undefined) {
        return {
          kind: "error",
          error: err(
            "unsupported-syntax",
            "unterminated template literal escape",
          ),
        };
      }
      const ESC: Record<string, string> = {
        n: "\n",
        r: "\r",
        t: "\t",
        "\\": "\\",
        "`": "`",
        $: "$",
      };
      value += ESC[next] ?? next;
      i += 2;
      continue;
    }
    if (ch === "$" && src[i + 1] === "{") {
      return {
        kind: "error",
        error: err(
          "unsupported-syntax",
          "template literal interpolation (`${...}`) is not supported — use string concatenation downstream",
        ),
      };
    }
    if (ch === "`") {
      return { kind: "ok", value, next: i + 1 };
    }
    value += ch;
    i++;
  }
  return {
    kind: "error",
    error: err("unsupported-syntax", "unterminated template literal"),
  };
}

function readNumber(
  src: string,
  start: number,
): { value: number; next: number } {
  let j = start;
  if (src[j] === "-") j++;
  while (j < src.length && /[0-9]/.test(src[j]!)) j++;
  if (src[j] === ".") {
    j++;
    while (j < src.length && /[0-9]/.test(src[j]!)) j++;
  }
  if (src[j] === "e" || src[j] === "E") {
    j++;
    if (src[j] === "+" || src[j] === "-") j++;
    while (j < src.length && /[0-9]/.test(src[j]!)) j++;
  }
  return { value: Number(src.slice(start, j)), next: j };
}

// ---------------------------------------------------------------------------
// Internal — token-stream + helpers
// ---------------------------------------------------------------------------

export class TokenStream {
  private readonly tokens: readonly Token[];
  private idx = 0;
  constructor(tokens: readonly Token[]) {
    this.tokens = tokens;
  }
  peek(): Token | undefined {
    return this.tokens[this.idx];
  }
  peekAt(offset: number): Token | undefined {
    return this.tokens[this.idx + offset];
  }
  next(): Token | undefined {
    return this.tokens[this.idx++];
  }
  atEnd(): boolean {
    return this.idx >= this.tokens.length;
  }
  consumePunct(value: string): boolean {
    const tok = this.peek();
    if (!tok || tok.kind !== "punct" || tok.value !== value) return false;
    this.idx++;
    return true;
  }
  consumeIdent(value: string): boolean {
    const tok = this.peek();
    if (!tok || tok.kind !== "ident" || tok.value !== value) return false;
    this.idx++;
    return true;
  }
}

export function topLevelSemicolons(tokens: readonly Token[]): number[] {
  const indices: number[] = [];
  let depth = 0;
  for (let i = 0; i < tokens.length; i++) {
    const tok = tokens[i]!;
    if (tok.kind !== "punct") continue;
    if (tok.value === "(" || tok.value === "[" || tok.value === "{") {
      depth++;
    } else if (tok.value === ")" || tok.value === "]" || tok.value === "}") {
      depth--;
    } else if (tok.value === ";" && depth === 0) {
      indices.push(i);
    }
  }
  return indices;
}

// ---------------------------------------------------------------------------
// Internal — small helpers shared by lexer + parser
// ---------------------------------------------------------------------------

export function describeToken(tok: Token): string {
  if (tok.kind === "string") return `"${tok.value}"`;
  return String(tok.value);
}

// Sprint 383 (sprint-382 review note 1) — single error helper. The previous
// `errMake`/`errRaw` pair was a no-op duplicate that obscured intent.
export function err(
  errorKind: MongoshErrorKind,
  message: string,
): MongoshParseError {
  return { kind: "error", errorKind, message };
}

export function tokErr(message: string): TokenizeErr {
  return { kind: "error", error: err("unsupported-syntax", message) };
}

File: src-tauri/mongosh-parser-core/src/lexer.rs
//! Hand-written character-level lexer for the sprint-401 mongosh grammar
//! slice. Mirrors `src/lib/mongo/mongoshAst/lexer.ts` (sprint-384) one-for-
//! one — same tokens, same comment / template / string rules, same head-
//! keyword sniff. Behavior parity is verified by `mongoshAst.test.ts` on
//! the frontend side.
//!
//! Design:
//! - No regex / nom / logos — hand-rolled keeps the WASM bundle minimal.
//! - Returns `Vec<Token>` or a `LexError` (a `MongoshStatement::Error`
//!   variant). Never panics on user-input paths.
//! - Identifier semantics are the JS-ish union of `[A-Za-z_$][A-Za-z0-9_$]*`
//!   — `$`-prefixed idents are valid (mongosh uses `$sum`, `$set`, ...).

use crate::ast::MongoshErrorKind;

#[derive(Debug, Clone, PartialEq)]
pub enum Token {
    Ident(String),
    StringLit(String),
    NumberLit(f64),
    /// Single-char punctuation: `{` `}` `[` `]` `(` `)` `,` `:` `;` `.`
    Punct(char),
}

#[derive(Debug, Clone, PartialEq)]
pub struct Spanned {
    pub token: Token,
    /// 0-based byte offset where this token starts.
    pub at: usize,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LexError {
    pub error_kind: MongoshErrorKind,
    pub message: String,
}

const PUNCT_CHARS: &[char] = &['{', '}', '[', ']', '(', ')', ',', ':', ';', '.'];

/// Lex the input into a flat token stream. Whitespace and comments are
/// stripped; the parser never sees them.
pub fn lex(input: &str) -> Result<Vec<Spanned>, LexError> {
    // Operate on `chars` so multi-byte UTF-8 in string literals (`"héllo"`)
    // passes through cleanly. The vast majority of token boundaries are
    // single ASCII chars so the conversion cost is negligible.
    let chars: Vec<char> = input.chars().collect();
    let mut tokens: Vec<Spanned> = Vec::new();
    let mut i = 0usize;

    // Pre-compute char→byte offsets so spans match the original `&str`. The
    // TS side reports byte offsets too; matching keeps error messages
    // grep-able across the boundary.
    let byte_offsets = char_byte_offsets(input, chars.len());

    while i < chars.len() {
        let ch = chars[i];
        let at = byte_offsets[i];

        // Whitespace.
        if matches!(ch, ' ' | '\t' | '\n' | '\r') {
            i += 1;
            continue;
        }

        // Line comment `// ... \n`.
        if ch == '/' && chars.get(i + 1) == Some(&'/') {
            while i < chars.len() && chars[i] != '\n' {
                i += 1;
            }
            continue;
        }

        // Block comment `/* ... */` — non-nested (mongosh parity: the first
        // `*/` closes; any trailing junk surfaces as a parse error later).
        if ch == '/' && chars.get(i + 1) == Some(&'*') {
            i += 2;
            while i < chars.len() && !(chars[i] == '*' && chars.get(i + 1) == Some(&'/')) {
                i += 1;
            }
            if i >= chars.len() {
                return Err(lex_err("unterminated /* ... */ comment"));
            }
            i += 2;
            continue;
        }

        // Arrow function `=>` — explicit reject (sprint-382 invariant).
        if ch == '=' && chars.get(i + 1) == Some(&'>') {
            return Err(lex_err(
                "arrow functions (`=>`) are not supported \
                 — callbacks are outside scope",
            ));
        }

        // Template literal `` `...` `` — interpolation-free → string.
        if ch == '`' {
            let (value, consumed) = read_template(&chars, i)?;
            tokens.push(Spanned {
                token: Token::StringLit(value),
                at,
            });
            i += consumed;
            continue;
        }

        // Single- or double-quoted string.
        if ch == '"' || ch == '\'' {
            let (value, consumed) = read_string(&chars, i, ch)?;
            tokens.push(Spanned {
                token: Token::StringLit(value),
                at,
            });
            i += consumed;
            continue;
        }

        // Number literal — leading `-` only if followed by a digit, matches
        // TS lexer's branch.
        if ch == '-'
            && chars
                .get(i + 1)
                .map(|c| c.is_ascii_digit())
                .unwrap_or(false)
        {
            let (value, consumed) = read_number(&chars, i)?;
            tokens.push(Spanned {
                token: Token::NumberLit(value),
                at,
            });
            i += consumed;
            continue;
        }
        if ch.is_ascii_digit() {
            let (value, consumed) = read_number(&chars, i)?;
            tokens.push(Spanned {
                token: Token::NumberLit(value),
                at,
            });
            i += consumed;
            continue;
        }

        // Identifier (incl. `$`-prefixed).
        if is_ident_start(ch) {
            let mut j = i + 1;
            while j < chars.len() && is_ident_continue(chars[j]) {
                j += 1;
            }
            let name: String = chars[i..j].iter().collect();
            tokens.push(Spanned {
                token: Token::Ident(name),
                at,
            });
            i = j;
            continue;
        }

        if PUNCT_CHARS.contains(&ch) {
            tokens.push(Spanned {
                token: Token::Punct(ch),
                at,
            });
            i += 1;
            continue;
        }

        return Err(lex_err(&format!(
            "unexpected character `{}` at offset {}",
            ch, at
        )));
    }

    Ok(tokens)
}

fn read_string(chars: &[char], start: usize, quote: char) -> Result<(String, usize), LexError> {
    let mut i = start + 1;
    let mut out = String::new();
    while i < chars.len() {
        let ch = chars[i];
        if ch == '\\' {
            let next = match chars.get(i + 1) {
                Some(c) => *c,
                None => return Err(lex_err("unterminated string escape")),
            };
            if next == 'u' {
                // \uXXXX — exactly 4 hex digits.
                if i + 5 >= chars.len() {
                    return Err(lex_err("invalid \\u escape in string"));
                }
                let hex: String = chars[i + 2..i + 6].iter().collect();
                if !hex.chars().all(|c| c.is_ascii_hexdigit()) || hex.len() != 4 {
                    return Err(lex_err("invalid \\u escape in string"));
                }
                let code = u32::from_str_radix(&hex, 16)
                    .map_err(|_| lex_err("invalid \\u escape in string"))?;
                if let Some(c) = char::from_u32(code) {
                    out.push(c);
                } else {
                    // TS `String.fromCharCode` accepts surrogates; we accept
                    // by emitting `\u{FFFD}` (replacement) for unpaired
                    // surrogates so the lexer never errors on input the TS
                    // side would have accepted.
                    out.push('\u{FFFD}');
                }
                i += 6;
                continue;
            }
            let mapped = match next {
                'n' => '\n',
                'r' => '\r',
                't' => '\t',
                'b' => '\u{0008}',
                'f' => '\u{000C}',
                '\\' => '\\',
                '\'' => '\'',
                '"' => '"',
                '/' => '/',
                other => other,
            };
            out.push(mapped);
            i += 2;
            continue;
        }
        if ch == quote {
            return Ok((out, i + 1 - start));
        }
        out.push(ch);
        i += 1;
    }
    Err(lex_err("unterminated string literal"))
}

fn read_template(chars: &[char], start: usize) -> Result<(String, usize), LexError> {
    let mut i = start + 1;
    let mut out = String::new();
    while i < chars.len() {
        let ch = chars[i];
        if ch == '\\' {
            let next = match chars.get(i + 1) {
                Some(c) => *c,
                None => return Err(lex_err("unterminated template literal escape")),
            };
            let mapped = match next {
                'n' => '\n',
                'r' => '\r',
                't' => '\t',
                '\\' => '\\',
                '`' => '`',
                '$' => '$',
                other => other,
            };
            out.push(mapped);
            i += 2;
            continue;
        }
        if ch == '$' && chars.get(i + 1) == Some(&'{') {
            return Err(lex_err(
                "template literal interpolation (`${...}`) is not supported \
                 — use string concatenation downstream",
            ));
        }
        if ch == '`' {
            return Ok((out, i + 1 - start));
        }
        out.push(ch);
        i += 1;
    }
    Err(lex_err("unterminated template literal"))
}

fn read_number(chars: &[char], start: usize) -> Result<(f64, usize), LexError> {
    let mut j = start;
    if chars[j] == '-' {
        j += 1;
    }
    while j < chars.len() && chars[j].is_ascii_digit() {
        j += 1;
    }
    if chars.get(j) == Some(&'.') {
        j += 1;
        while j < chars.len() && chars[j].is_ascii_digit() {
            j += 1;
        }
    }
    if let Some(&e) = chars.get(j) {
        if e == 'e' || e == 'E' {
            j += 1;
            if let Some(&sign) = chars.get(j) {
                if sign == '+' || sign == '-' {
                    j += 1;
                }
            }
            while j < chars.len() && chars[j].is_ascii_digit() {
                j += 1;
            }
        }
    }
    let slice: String = chars[start..j].iter().collect();
    let value = slice
        .parse::<f64>()
        .map_err(|_| lex_err("internal: invalid numeric literal"))?;
    Ok((value, j - start))
}

fn is_ident_start(ch: char) -> bool {
    ch.is_ascii_alphabetic() || ch == '_' || ch == '$'
}

fn is_ident_continue(ch: char) -> bool {
    ch.is_ascii_alphanumeric() || ch == '_' || ch == '$'
}

fn char_byte_offsets(input: &str, n_chars: usize) -> Vec<usize> {
    // Map char-index → byte-offset so we can report TS-comparable spans.
    // Allocates `n_chars + 1` to allow a one-past-end probe.
    let mut offsets = Vec::with_capacity(n_chars + 1);
    for (b, _) in input.char_indices() {
        offsets.push(b);
    }
    offsets.push(input.len());
    offsets
}

fn lex_err(msg: &str) -> LexError {
    LexError {
        error_kind: MongoshErrorKind::UnsupportedSyntax,
        message: msg.to_string(),
    }
}

// ---------------------------------------------------------------------------
// Top-level semicolon scan + identifier-only sniff helpers used by parser.
// ---------------------------------------------------------------------------

/// Returns the *indices* (into the token vector) of every top-level `;` —
/// those that appear at brace/bracket/paren depth 0. The parser uses this
/// to decide whether the input is a single statement or a multi-statement
/// chain.
pub fn top_level_semicolons(tokens: &[Spanned]) -> Vec<usize> {
    let mut out = Vec::new();
    let mut depth: i32 = 0;
    for (i, sp) in tokens.iter().enumerate() {
        match &sp.token {
            Token::Punct('(') | Token::Punct('[') | Token::Punct('{') => depth += 1,
            Token::Punct(')') | Token::Punct(']') | Token::Punct('}') => depth -= 1,
            Token::Punct(';') if depth == 0 => out.push(i),
            _ => {}
        }
    }
    out
}

/// Stringify a token for inclusion in user-visible error messages — matches
/// the TS `describeToken` helper.
pub fn describe_token(tok: &Token) -> String {
    match tok {
        Token::StringLit(s) => format!("\"{}\"", s),
        Token::NumberLit(n) => format_number_for_message(*n),
        Token::Ident(s) => s.clone(),
        Token::Punct(c) => c.to_string(),
    }
}

fn format_number_for_message(n: f64) -> String {
    // Match JS's `String(n)` behavior for integer-valued floats: emit `1`
    // not `1.0`. Anything else falls back to Rust's default.
    if n.is_finite() && n.fract() == 0.0 && n.abs() < 1e16 {
        format!("{}", n as i64)
    } else {
        format!("{}", n)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn lex_tokens(input: &str) -> Vec<Token> {
        lex(input)
            .expect("lex")
            .into_iter()
            .map(|s| s.token)
            .collect()
    }

    #[test]
    fn ac_l1_idents_strings_numbers_puncts() {
        assert_eq!(lex_tokens("db"), vec![Token::Ident("db".into())]);
        assert_eq!(
            lex_tokens("\"hello\""),
            vec![Token::StringLit("hello".into())]
        );
        assert_eq!(lex_tokens("'x'"), vec![Token::StringLit("x".into())]);
        assert_eq!(lex_tokens("42"), vec![Token::NumberLit(42.0)]);
        assert_eq!(lex_tokens("-3"), vec![Token::NumberLit(-3.0)]);
        assert_eq!(lex_tokens("1.5"), vec![Token::NumberLit(1.5)]);
        assert_eq!(lex_tokens("2e3"), vec![Token::NumberLit(2000.0)]);
        assert_eq!(lex_tokens("("), vec![Token::Punct('(')]);
    }

    #[test]
    fn ac_l2_line_comment_stripped() {
        let toks = lex_tokens("// hi\ndb");
        assert_eq!(toks, vec![Token::Ident("db".into())]);
    }

    #[test]
    fn ac_l3_block_comment_stripped() {
        let toks = lex_tokens("/* hi */db");
        assert_eq!(toks, vec![Token::Ident("db".into())]);
    }

    #[test]
    fn ac_l3_block_comment_unterminated_errors() {
        let err = lex("/* never closed").unwrap_err();
        assert_eq!(err.error_kind, MongoshErrorKind::UnsupportedSyntax);
        assert!(err.message.contains("unterminated"));
    }

    #[test]
    fn ac_l4_template_interpolation_free_is_string() {
        assert_eq!(
            lex_tokens("`hello`"),
            vec![Token::StringLit("hello".into())]
        );
    }

    #[test]
    fn ac_l5_template_interpolation_rejected() {
        let err = lex("`hello ${x}`").unwrap_err();
        assert!(err.message.contains("interpolation"));
    }

    #[test]
    fn ac_l6_arrow_rejected() {
        let err = lex("d => d.name").unwrap_err();
        assert!(err.message.contains("arrow"));
    }

    #[test]
    fn ac_l7_unterminated_string_errors() {
        let err = lex("\"never closed").unwrap_err();
        assert!(err.message.contains("unterminated"));
    }

    #[test]
    fn dollar_idents_ok() {
        assert_eq!(lex_tokens("$sum"), vec![Token::Ident("$sum".into())]);
    }

    #[test]
    fn unicode_escape_in_string() {
        assert_eq!(
            lex_tokens("\"\\u0041B\""),
            vec![Token::StringLit("AB".into())]
        );
    }

    #[test]
    fn top_level_semicolons_finds_correct_indices() {
        let tokens = lex("a; b").unwrap();
        let semis = top_level_semicolons(&tokens);
        assert_eq!(semis, vec![1]);
    }

    #[test]
    fn nested_semicolons_ignored() {
        let tokens = lex("[1;2]").unwrap();
        let semis = top_level_semicolons(&tokens);
        assert!(semis.is_empty());
    }
}
